sklearn/tests/test_public_functions.py

```diff
@@ -0,0 +1,77 @@
+from importlib import import_module
+from inspect import signature
+
+import pytest
+
+from sklearn.utils._param_validation import generate_invalid_param_val
+from sklearn.utils._param_validation import generate_valid_param
+from sklearn.utils._param_validation import make_constraint
+from sklearn.cluster import kmeans_plusplus
+
+
+PARAM_VALIDATION_FUNCTION_LIST = [kmeans_plusplus]
+
+
+@pytest.mark.parametrize("func", PARAM_VALIDATION_FUNCTION_LIST)
+def test_function_param_validation(func):
+    """Check that an informative error is raised when the value of a parameter does not
+    have an appropriate type or value.
+
+    the parameter constraints for the function must be named
+    _parameter_constraints_of_<func_name> and be located in the same module.
+    """
+    func_sig = signature(func)
+    func_params = [
+        p.name
+        for p in func_sig.parameters.values()
+        if p.kind not in (p.VAR_POSITIONAL, p.VAR_KEYWORD)
+    ]
+    func_module = import_module(func.__module__)
+    func_name = func.__name__
+    parameter_constraints = getattr(
+        func_module, f"_parameter_constraints_of_{func_name}"
+    )
+
+    # generate valid values for the required parameters
+    required_params = [
+        p.name for p in func_sig.parameters.values() if p.default is p.empty
+    ]
+    required_params = {
+        p: generate_valid_param(make_constraint(parameter_constraints[p][0]))
+        for p in required_params
+    }
+
+    # check that there is a constraint for each parameter
+    if func_params:
+        err_msg = (
+            f"Mismatch between _parameter_constraints_of_{func_name} and the parameters"
+            f" of {func_name}."
+        )
+        assert list(parameter_constraints.keys()) == func_params, err_msg
```

Comment:
Could we generate the parameter diff in the `err_msg`?

Reply:
I think it's already pretty clear:
```
E           AssertionError: Mismatch between _parameter_constraints_of_kmeans_plusplus and the parameters of kmeans_plusplus.
E           assert ['X', 'n_clus...rials', 'lul'] == ['X', 'n_clus...local_trials']
E             Left contains one more item: 'lul'
E             Full diff:
E             - ['X', 'n_clusters', 'x_squared_norms', 'random_state', 'n_local_trials']
E             + ['X', 'n_clusters', 'x_squared_norms', 'random_state', 'n_local_trials', 'lul']
E             ?                                                                        +++++++
```

Reply:
Right. I did not think the assert was doing something that clear :)